src/main/kotlin/com/avail/optimizer/OptimizationPhase.kt

```diff
@@ -0,0 +1,305 @@
+	 * Transform each reference to an old register into a reference to the
+	 * replacement, updating structures as needed.
+	 */
+	REPLACE_REGISTERS_BY_COLOR({ it.replaceRegistersByColor() }),
+
+	/**
+	 * Remove any remaining moves between two registers of the same color.
+	 */
+	REMOVE_SAME_COLOR_MOVES({ it.removeSameColorMoves() }),
+
+	/**
+	 * Every L2PcOperand that leads to an L2_JUMP should now be redirected
+	 * to the target of the jump (transitively, if the jump leads to another
+	 * jump).  We specifically do this after inserting phi moves to ensure
+	 * we don't jump past irremovable phi moves.
+	 */
+	ADJUST_EDGES_LEADING_TO_JUMPS({ it.adjustEdgesLeadingToJumps() }),
+
+	/**
+	 * Having adjusted edges to avoid landing on L2_JUMPs, some blocks may
+	 * have become unreachable.
+	 */
+	REMOVE_UNREACHABLE_BLOCKS({ it.removeUnreachableBlocks() }),
+
+	/**
+	 * Choose an order for the blocks.  This isn't important while we're
+	 * interpreting L2Chunks, but it will ultimately affect the quality of
+	 * JVM translation.  Prefer to have the target block of an unconditional
+	 * jump to follow the jump, since final code generation elides the jump.
+	 */
+	ORDER_BLOCKS({ it.orderBlocks() });
+	// Additional optimization ideas:
+	//    -Strengthen the types of all registers and register uses.
+	//    -Ask instructions to regenerate if they want.
+	//    -When optimizing, keep track of when a TypeRestriction on a phi
+	//     register is too weak to qualify, but the types of some of the phi
+	//     source registers would qualify it for a reasonable expectation of
+	//     better performance.  Write a hint into such phis.  If we have a
+	//     high enough requested optimization level, apply code-splitting.
+	//     The block that defines that phi can be duplicated for each
+	//     interesting incoming edge.  That way the duplicated blocks will
+	//     get more specific types to work with.
+	//    -Splitting for int32s.
+	//    -Leverage more inter-primitive identities.
+	//    -JVM target.
+
+	/** The [Statistic] for tracking this pass's cost.  */
+	val stat: Statistic = Statistic(
+		name, StatisticReport.L2_OPTIMIZATION_TIME)
+
+	/** The [StateFlag]s to require to already be set as preconditions.  */
+	val requiresFlags =
+		mutableListOf<KClass<out StateFlag>>()
+
+	/** The [StateFlag]s that should already be clear as preconditions.  */
+	val requiresNotFlags=
+		mutableListOf<KClass<out StateFlag>>()
+
+	/** The [StateFlag]s to set after this phase.  */
+	val setsFlags =
+		mutableListOf<KClass<out StateFlag>>()
+
+	/** The [StateFlag]s to clear after this phase.  */
+	val clearsFlags =
+		mutableListOf<KClass<out StateFlag>>()
+
+	/**
+	 * Create the enumeration value, capturing a parameter to pass to the
+	 * optimization lambda.
+	 *
+	 * @param action
+	 *   The action to perform.
+	 * @param value
+	 *   The actual value to pass to the action.
+	 */
+	constructor(
+		action: (L2Optimizer, Any) -> Unit,
```

Comment:
It would probably be more elegant here to use an extension function:

```kt
action: L2Optimizer.(Any) -> Unit
```

Then you could write an `enum` case something like:

```kt
REMOVE_UNREACHABLE_BLOCKS({ removeUnreachableBlocks() })
```